src/addr_validate.rs

```diff
@@ -42,56 +54,50 @@ fn create_pipe() -> nix::Result<(i32, i32)> {
 }
 
 fn open_pipe() -> nix::Result<()> {
-    MEM_VALIDATE_PIPE.with(|pipes| {
-        let mut pipes = pipes.borrow_mut();
+    let pipes = unsafe { &mut *MEM_VALIDATE_PIPE.get() };
 
-        // ignore the result
-        let _ = close(pipes[0]);
-        let _ = close(pipes[1]);
+    // ignore the result
+    let _ = close(pipes[0]);
+    let _ = close(pipes[1]);
 
-        let (read_fd, write_fd) = create_pipe()?;
+    let (read_fd, write_fd) = create_pipe()?;
 
-        pipes[0] = read_fd;
-        pipes[1] = write_fd;
+    pipes[0] = read_fd;
+    pipes[1] = write_fd;
 
-        Ok(())
-    })
+    Ok(())
 }
 
 pub fn validate(addr: *const libc::c_void) -> bool {
     const CHECK_LENGTH: usize = 2 * size_of::<*const libc::c_void>() / size_of::<u8>();
 
     // read data in the pipe
-    let valid_read = MEM_VALIDATE_PIPE.with(|pipes| {
-        let pipes = pipes.borrow();
-        loop {
-            let mut buf = [0u8; CHECK_LENGTH];
-
-            match read(pipes[0], &mut buf) {
-                Ok(bytes) => break bytes > 0,
-                Err(_err @ Errno::EINTR) => continue,
-                Err(_err @ Errno::EAGAIN) => break true,
-                Err(_) => break false,
-            }
+    let pipes = unsafe { &*MEM_VALIDATE_PIPE.get() };
```

Comment:
BTW, could you explain why we can use a pipe to validate the addr?

Reply:
FYI https://github.com/tikv/pprof-rs/pull/116#issuecomment-1092388824

Reply:
I mean better to add some comments